File: app/src/main/java/com/leodeleon/popmovies/data/MovieAPI.java
package com.leodeleon.popmovies.data;

import com.leodeleon.popmovies.model.MovieDetail;
import com.leodeleon.popmovies.model.MovieResponse;
import com.leodeleon.popmovies.model.VideoResponse;
import io.reactivex.Single;
import retrofit2.Retrofit;
import retrofit2.http.GET;
import retrofit2.http.Path;
import retrofit2.http.Query;

public class MovieAPI {
  private MovieService movieService;

  public MovieAPI(Retrofit retrofit) {
    movieService = retrofit.create(MovieService.class);
  }

  public Single<MovieResponse> getPopularMovies(int page) {
    return movieService.getPopularMovies(page);
  }

  public Single<MovieResponse> getTopRatedMovies(int page) {
    return movieService.getTopRatedMovies(page);
  }

  public Single<MovieDetail> getMovieDetail(int id) {
    return movieService.getMovie(id);
  }

  public Single<VideoResponse> getVideos(int id) {
    return movieService.getVideos(id);
  }


  interface MovieService {
    @GET("movie/popular") Single<MovieResponse> getPopularMovies(@Query("page") int page);

    @GET("movie/top_rated") Single<MovieResponse> getTopRatedMovies(@Query("page") int page);

    @GET("movie/{movieId}") Single<MovieDetail> getMovie(@Path("movieId") int movieId);

    @GET("movie/{movieId}/videos") Single<VideoResponse> getVideos(@Path("movieId") int movieId);
  }
}

File: app/src/main/java/com/leodeleon/popmovies/data/remote/MovieAPI.kt
package com.leodeleon.popmovies.data.remote

import com.leodeleon.popmovies.model.MovieDetail
import com.leodeleon.popmovies.model.MovieResponse
import com.leodeleon.popmovies.model.VideoResponse
import io.reactivex.Single
import retrofit2.Retrofit
import retrofit2.http.GET
import retrofit2.http.Path
import retrofit2.http.Query

class MovieAPI(val movieService: MovieService) {

  fun getPopularMovies(page: Int): Single<MovieResponse> {
    return movieService.getPopularMovies(page)
  }

  fun getTopRatedMovies(page: Int): Single<MovieResponse> {
    return movieService.getTopRatedMovies(page)
  }

  fun getMovieDetail(id: Int): Single<MovieDetail> {
    return movieService.getMovie(id)
  }

  fun getVideos(id: Int): Single<VideoResponse> {
    return movieService.getVideos(id)
  }


  interface MovieService {
    @GET("movie/popular") fun getPopularMovies(@Query("page") page: Int): Single<MovieResponse>

    @GET("movie/top_rated") fun getTopRatedMovies(@Query("page") page: Int): Single<MovieResponse>

    @GET("movie/{movieId}") fun getMovie(@Path("movieId") movieId: Int): Single<MovieDetail>

    @GET("movie/{movieId}/videos") fun getVideos(@Path("movieId") movieId: Int): Single<VideoResponse>
  }
}
